### ISSUE TITLE
Clang-CL 15.x reporting warning when using /clang:-xc

### ISSUE THREAD

**Issue by dgehriger** [llvm/llvm-project#59307]
OS: Windows

Unlike previous versions, when running clang-cl.exe version 15.x as follows from a cmd.exe prompt:

```ps
clang-cl /WX /clang:-xc /clang:-E /clang:-dM nul
```

It outputs the following warning:

```console
clang-cl: error: '-x c' after last input file has no effect [-Werror,-Wunused-command-line-argument]
```

Expected outcome is no error, and dumping of preprocessor variables.

Among other things, this prevents clang-cl 15.x from being used by Visual Studio Code's C++ extension for Intellisense.

**Comment by efriedma-quic** [comment]
As a workaround, the following should work on both old and new versions of clang-cl:

```
clang-cl /WX /TC /E /clang:-dM nul
```

**Comment by AaronBallman** [comment]
@Fznamznon -- I assigned this one to you because I figured it was one you might be able to knock out pretty quickly, but if you don't feel like it, feel free to unassign yourself. :-)

**Comment by Fznamznon** [comment]
Ah, I see. In cl mode `/clang:` arguments are parsed and appended at the end of the list. I would suggest just to not emit a warning in cl mode since it was implemented to match gcc behavior anyway (e2a1f8ec27b88be670cd867c43588f24516d2bbf).
Another thing comes in mind that if all `clang:` arguments go to the end of list, `clang:-x` doesn't make sense at all, which is actually we get warning about just not with the right works.

**Comment by AaronBallman** [comment]
@MaskRay -- is it expected that all `/clang:` arguments are appended? That seems likely to cause a fair amount of confusion in practice, but perhaps I'm missing something.

**Comment by Fznamznon** [comment]
In case silencing the warning in cl mode sounds okay-ish, I've prepared https://reviews.llvm.org/D142757 . Otherwise if we want to emit a correct warning in cl mode, we need to change how to `/clang:` arguments are processed and for that I'm not sure I have enough driver experience to do quickly.

**Comment by MaskRay** [comment]
The error is emitted as intended. `/clang:-xc nul` is handled like `nul -xc` and the `-xc` does not take effect.
This likely has been the case since https://reviews.llvm.org/D53457 added `/clang:` in 2018-11, therefore I don't think the new feature has a legitimate use case.

So simply ignoring the error in CL mode (https://reviews.llvm.org/D142757) looks wrong to me.
I think we should just report an error for `/clang:-xc`. Left a comment on the patch about the change.

**Comment by huangqinjin** [comment]
`clang-cl -help` shows that `-x` is supported directly. But `clang-cl -x c++-module` report an error 
```
unsupported option '-x c++-module'; did you mean '/TC' or '/TP'?
```

Unfortunately there is no coresponding flag to specify C++20 Modules.
